### PR TITLE
Mesos/Marathon support for latent worker

### PR DESCRIPTION
**Pull request by tardyp**
This adds a bit more docker factorization, and implement a simple support for Marathon.

This has not been tested a lot yet beyond the integration tests, but it managed to create 100 containers in parallel via the integration tests.

See the doc for more information

### CODE REVIEW

File: master/docs/manual/cfg-workers-docker.rst [hunk 1]
```diff
@@ -307,3 +307,62 @@ In addition to the arguments available for any :ref:`Latent-Workers`, :class:`Hy
 .. _Hyper: https://hyper.sh
 .. _HyperPricing: https://hyper.sh/pricing.html
 .. _DockerHub: https://hub.docker.com/
+
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-docker.rst]: [nit] i'd keep just one empty line here

File: master/docs/manual/cfg-workers-docker.rst [hunk 2]
```diff
@@ -307,3 +307,62 @@ In addition to the arguments available for any :ref:`Latent-Workers`, :class:`Hy
...
+
+
+Marathon latent worker
+======================
+
+Marathon_ Marathon is a production-grade container orchestration platform for Mesosphere's Data-center Operating System (DC/OS) and Apache ``Mesos``.
+
+Buildbot supports using Marathon_ to host your latent workers.
+It requires either `txrequests`_ or `treq`_ to be installed to allow interaction with http server.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-docker.rst]: [nit] can we just refer to the section that talks about `HTTPClientService`?

File: master/docs/manual/cfg-workers-docker.rst [hunk 3]
```diff
@@ -307,3 +307,62 @@ In addition to the arguments available for any :ref:`Latent-Workers`, :class:`Hy
...
+
+.. @cindex MarathonLatentWorker
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-docker.rst]: sphinx does not support this kind of index, this should be removed.

File: master/docs/manual/cfg-workers-docker.rst [hunk 4]
```diff
@@ -307,3 +307,62 @@ In addition to the arguments available for any :ref:`Latent-Workers`, :class:`Hy
...
+.. py:class:: buildbot.worker.marathon.MarathonLatentWorker
+.. py:class:: buildbot.plugins.worker.MarathonLatentWorker
+
+Using the Marathon latent worker, an attempt is made at instantiating a fresh image upon each build, assuring consistency of the environment between builds.
```
Comment (sa2ajj) [inline master/docs/manual/cfg-workers-docker.rst]: > Marathon latent worker attempts to instantiate a build using a fresh image to assure consistency...

?

Comment (seankelly) [inline master/docs/manual/cfg-workers-docker.rst]: How about,
> The Marathon latent worker attempts to instantiate a fresh image for each build to assure consistency of the environment between builds.

That raises the question of what happens if the attempt fails.

**Comment by tardyp** [comment]
@sa2ajj I think I addressed the concerns. Let me know if I forgot something.

File: master/buildbot/steps/source/git.py [hunk 5]
```diff
@@ -140,7 +140,7 @@ def __init__(self, repourl=None, branch='HEAD', mode='incremental', method=None,
         self.srcdir = 'source'
         self.origin = origin
         Source.__init__(self, **kwargs)
-
+        print "shallow:", shallow
```
Comment (sa2ajj) [inline master/buildbot/steps/source/git.py]: this is probably a stray debug print

File: master/buildbot/test/integration/test_marathon.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+from buildbot.config import BuilderConfig
+from buildbot.plugins import schedulers
+from buildbot.plugins import steps
+from buildbot.process.factory import BuildFactory
+from buildbot.process.results import SUCCESS
+from buildbot.test.util.integration import RunMasterBase
+from buildbot.worker.marathon import MarathonLatentWorker
+
+# This integration test creates a master and marathon worker environment,
```
Comment (sa2ajj) [inline master/buildbot/test/integration/test_marathon.py]: :+1:

File: master/buildbot/util/__init__.py [hunk 7]
```diff
@@ -415,6 +415,18 @@ def needs_wrapping(line):
     return wrapped_text
 
 
+def dictionary_merge(a, b):
+    """merges dictionary b into a
+       Like dict.update, but recursive
+    """
+    for key, value in b.items():
+        if key in a and isinstance(a[key], dict) and isinstance(value, dict):
+                dictionary_merge(a[key], b[key])
```
Comment (sa2ajj) [inline master/buildbot/util/__init__.py]: the indentation seems to be off here

**Comment by sa2ajj** [comment]
the indentation issue (which I'd expect our tools would catch) is pretty much the issue now

**Comment by tardyp** [comment]
updated

**Comment by sa2ajj** [comment]
rel notes seem to be the thing now :/

**Comment by Frodox** [comment]
Cool :+1: 
Any plans for k8s support? :)

**Pull request merged**